runtime/src/accounts_db.rs

```diff
@@ -629,18 +629,55 @@ impl AccountsDB {
         false
     }
 
-    // Purge zero lamport accounts for garbage collection purposes
+    // Reclaim older states of rooted non-zero lamport accounts as a general
+    // AccountsDB bloat mitigation and preprocess for better zero-lamport purging.
+    fn clean_old_rooted_accounts(&self, all_pubkeys: Vec<Pubkey>) {
+        let mut measure = Measure::start("clean_old_root-ms");
+        let reclaim_vecs = all_pubkeys.par_chunks(4096).map(|pubkeys: &[Pubkey]| {
+            let mut reclaims = Vec::new();
+            let accounts_index = self.accounts_index.read().unwrap();
+            for pubkey in pubkeys {
+                accounts_index.clean_rooted_entries(&pubkey, &mut reclaims);
+            }
+            reclaims
+        });
+        let reclaims: Vec<_> = reclaim_vecs.flatten().collect();
+        measure.stop();
+        inc_new_counter_info!("clean-old-root-par-clean-ms", measure.as_ms() as usize);
+
+        let mut measure = Measure::start("clean_old_root-ms");
+        self.handle_reclaims(&reclaims);
+        measure.stop();
+        inc_new_counter_info!("clean-old-root-reclaim-ms", measure.as_ms() as usize);
+
+        let mut accounts_index = self.accounts_index.write().unwrap();
+        accounts_index.uncleaned_roots.clear();
+        drop(accounts_index);
+    }
+
+    // Purge zero lamport accounts and older rooted account states as garbage
+    // collection
     // Only remove those accounts where the entire rooted history of the account
     // can be purged because there are no live append vecs in the ancestors
-    pub fn purge_zero_lamport_accounts(&self, ancestors: &HashMap<u64, usize>) {
+    pub fn clean_accounts(&self, _ancestors: &HashMap<u64, usize>) {
         self.report_store_stats();
         let mut purges = HashMap::new();
+        let mut rooted_purges = Vec::new();
         let accounts_index = self.accounts_index.read().unwrap();
-        accounts_index.scan_accounts(ancestors, |pubkey, (account_info, slot)| {
-            if account_info.lamports == 0 && accounts_index.is_root(slot) {
+        accounts_index.scan_accounts(&HashMap::new(), |pubkey, (account_info, slot)| {
```

Comment:
As was [the case](https://github.com/solana-labs/solana/pull/8436/files#diff-2099c5256db4eb5975c8834af38f6456L640), we only need rooted index entries. So let's simplify things.